### ISSUE TITLE
Ongoing Problems with the xp system

### ISSUE THREAD

**Issue by AdamantineAmmonite** [NordicGamerFE/Ocean-Outlast#19]
Version: Expert 48 HF 1

Description:
After the removal of the XP Bank the xp system is still broken.
Apparently this only helps players who start a new world, where gaining and spending xp is now possible without a chance to encounter the bug.
In my existing world the bug is still there, the only change is that the bank is gone.

Is there any way to repair this or do I have to give up on my hours of progress and start over again?

Did you add additional mods?
no

**Comment by NordicGamerFE** [comment]
try to use /xp command and see if you are able to repair anything with that

**Comment by AdamantineAmmonite** [comment]
I did try that, but it seems the xp system is permanently broken. I can't seem to add or substract any xp by any means. Even commands can't change the xp anymore. After giving me xp by using the code im still at 0 xp.

I wanted to test this further so i started a new game after the hotfix. With the bank removed the error shouldn't occur anymore. I thought.
After playing for some time, spending some xp on skills and dying a few times i realized that in my new game the xp system is now broken as well. 
I assume it's the same error as before. It's impossible to gain xp anymore and when i check the tab where i can spend xp on skills it shows the negative number again.
I can't tell you when or why exactly this happened, but removing the XP Bank doesn't even solve the issue for new games it seems.

Maybe the XP Bank in particular is not the problem but there general xp management system via Improvable Skills.

**Comment by AdamantineAmmonite** [comment]
I did another test on the matter.
Everything is okay as long as I don't have the book for the skill mod in my possession.
If I have the book, but haven't assigned any points to skills, everything is still okay.
As soon as I have invested points into a skill and I die, xp gets substracted.
I assume the game recognizes that I have invested xp in skills.
Therefore it thinks I have xp, when in fact my xp bar is empty because I spent those points.
So when I die, I probably lose the xp I spent, but in fact don't have anymore.
My xp goes to a certain negative number.
The next time i gain xp, the game tries to add the xp to the negative number and my xp counter instantly jumps to an insanely high negative number, because the game bugs out.

**Comment by NordicGamerFE** [comment]
i found that using an experience oblisk works

**Comment by AdamantineAmmonite** [comment]
The obelisk apparently works fine on his own.
If you're at negative xp due to the bug, and you didn't store any xp in the obelisk before the bug, you have no xp to withdraw and due to the bug you cannot gain xp to put into the obelisk. So a new or empty obelisk at this point becomes useless.

If you store xp in the obelisk before dying, the bug happens and you can withdraw xp from it.
It looks like your xp bar is fixed, but when you store the max amount you have in the obelisk again, you suddenly store almost 5.000 levels of experience in it. 
So in the background the xp system is still bugged it seems, it just looks like its fixed.

After withdrawing those levels and dying again my game crashed.

It still kinda looks like the skills mess up the whole xp mechanic at this point, now matter how you interact with it.

So far everyone who invests in skills and dies later on has a bugged game i guess.

**Comment by AdamantineAmmonite** [comment]
Update: Ran another test.

Started a new game. I completely left Improvable Skills alone. Didn't even get the book for the mod.
After some time I died. After that I couldn't gain xp anymore, even though I didn't touch the mod.

For testing purposes i got the skill book after that and died again. I didn't buy any skills.
The book was showing me negative xp numbers again.

At this point I'm really confused. Apparently there's a bigger problem here.
The xp system seemly gets bugged in some way I don't understand. It appears that the mod Improvable Skills just highlights the problem since you're able to clearly see your xp numbers when using the book.
I guess the mod itself is not the reason for the bug since it happens anyway, even if you don't actively use the mod(?)

**Comment by NordicGamerFE** [comment]
Thank you for the information i will be running tests later today to find the source of the issue :D

**Comment by stijndp** [comment]
For testing purposes you can reset faulty save files by resetting XpLevel, XpP and XpTotal to 0 in the level.dat.
Or if you just want to recover a faulty save.

For example in a backup I have
XpLevel 22
XpP 0,111111
XpTotal 687
[Values are normal.](https://minecraft.gamepedia.com/Experience#Leveling_up)

backup 2 days later
XpLevel 0
XpP -3,067834E+08
XpTotal 2147483647

My XpTotal is 2147483647 which is the maximum allowed total xp since it's the max value of an int32.
But ingame in the skill book UI the value displays as -2147482648. This isn't the minimum value of an int32 but 100 higher.
I think there is a mod incompatible with 1.12.2 or a mod implemented 1.13 experience commands forward compatibility with bugs. I think the problems started after dying but don't take that as reproduction steps.

I also have a problem with food since my foodExhaustionLevel is at -18338,78. Which means I never have to eat. Somewhere a mod is doing - instead of +.

**Comment by AdamantineAmmonite** [comment]
Yes, the maximum xp number was so oddly specific that i figured it had something to with integers.
I'm still pretty sure that dying somehow triggers the bug since it interacts with your experience.

In the very first game that i encountered the bug in, my hunger was acting strange as well. For a long time I didn't have to eat at all but I figured that was maybe intentional. I think my food meter constantly went back up to full after sleeping.

Some time later it normalized and in the last test (other world) it didn't act up at all. It went down with time as expected.

Edit: I got the same xp values in my level.dat file as stijndp.

Edit: Thanks for the hint. Changing the indicated lines in the level.dat does fix the isse. At least until i die the next time i assume, but at least i can salvage my game.
